### PR TITLE
feat(client): ts-migrate Settings component

### PR DESCRIPTION
**Pull request by Nirajn2311**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ref #42256 

Current Progress:
- [x] /client/src/components/settings/About.js
- [ ] /client/src/components/settings/Certification.test.js
- [x] /client/src/components/settings/DangerZone.js
- [x] /client/src/components/settings/DeleteModal.js
- [x] /client/src/components/settings/Email.js
- [x] /client/src/components/settings/Honesty.js
- [ ] /client/src/components/settings/Honesty.test.js
- [x] /client/src/components/settings/Internet.js
- [x] /client/src/components/settings/Portfolio.js
- [x] /client/src/components/settings/Privacy.js
- [x] /client/src/components/settings/ResetModal.js
- [x] /client/src/components/settings/SectionHeader.js
- [x] /client/src/components/settings/Theme.js
- [x] /client/src/components/settings/ToggleSetting.js
- [x] /client/src/components/settings/Username.js
- [x] Review changes once again
- [x] Changing the filename from `PascalCase` to `kebab-case` and ensuring that the imports in code are all still valid.

### CODE REVIEW

**Review by raisedadead — COMMENTED** [review]
Hey there,

Thank you for taking the time to contribute to the TypeScript migration. 

The main issue has been updated a lot over the last few days, and so we suggest you re-read it and update this pull request to match if you have the time.

The main goals for a PR are:
- [x] Changing the filename from `PascalCase` to `kebab-case` and ensuring that the imports in code are all still valid.
- [x] Changing the file extension to .tsx (if a file has React JSX syntax)
- [x] Fixing as many TypeScript and Eslint errors as you can while refactoring the code.

If you need any assistance, please reach out in our [contributing chat room](https://chat.freecodecamp.org/channel/contributors).

**Comment by Nirajn2311** [comment]
Yikes I missed out the part with file name format, will change them today and I'll finish this PR over the weekend.

File: client/src/components/settings/DeleteModal.tsx [hunk 1]
```diff
@@ -36,6 +38,8 @@ function DeleteModal(props) {
         <p>{t('settings.danger.delete-p1')}</p>
         <p>{t('settings.danger.delete-p2')}</p>
         <p>
+          {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
+          {/* @ts-ignore */}
           <Trans email={email} i18nKey='settings.danger.delete-p3'>
```
Comment (ShaunSHamilton) [inline client/src/components/settings/DeleteModal.tsx:43]:
```suggestion
          <Trans i18nKey='settings.danger.delete-p3'>
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
This might not be entirely correct, but from what I have seen, we are using the Trans component incorrectly, and do not need to pass the properties to the props. That is why TS is complaining.

File: client/src/components/settings/Portfolio.tsx [hunk 2]
```diff
@@ -112,25 +122,26 @@ class PortfolioSettings extends Component {
...
 
-  isFormValid = id => {
-    const { portfolio } = this.state;
-    const toValidate = find(portfolio, createFindById(id));
-    if (!toValidate) {
-      return false;
-    }
-    const { title, url, image, description } = toValidate;
+  // TODO: Check if this function is required or not
+  // isFormValid = id => {
```
Comment (Nirajn2311) [inline client/src/components/settings/Portfolio.tsx:126]: Is this function required as it isn't used anywhere and it looks similar to the function above it namely `isFormPristine`

File: client/src/components/settings/Portfolio.tsx [hunk 3]
```diff
@@ -43,16 +49,18 @@ function createEmptyPortfolio() {
...
 
-function createFindById(id) {
-  return p => p.id === id;
+function createFindById(id: string) {
+  return (p: PortfolioValues) => p.id === id;
 }
 
 const mockEvent = {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
   preventDefault() {}
```
Comment (Nirajn2311) [inline client/src/components/settings/Portfolio.tsx:58]: I'm confused with what the `{}` braces does in this line. Also this function is called below in line 110. I've pasted the respective function that calls it here.
```js
handleRemoveItem = (id: string) => {
    return this.setState(
      state => ({
        portfolio: state.portfolio.filter(p => p.id !== id)
      }),
      () => this.handleSubmit(mockEvent)
    );
  };
```

File: client/src/components/settings/DeleteModal.tsx (hunk 1 above)
Comment (Nirajn2311) [reply client/src/components/settings/DeleteModal.tsx:43]: I see. I haven't used `react-i18next` so wan't too sure what to do there and commented it out for later review. I have the last two files, the test files, remaining to be migrated. Also Oliver had given a link on using TS with `react-i18next`, so have to go through that. I had planned to go through the changes I did once I had migrated all the files incase I missed anything or did something the wrong way.  I'll correct this when I'm doing that review.

File: client/src/components/settings/Portfolio.tsx (hunk 3 above)
Comment (ShaunSHamilton) [reply client/src/components/settings/Portfolio.tsx:58]: I think this might be needed for testing, but I am unsure.

**Comment by ShaunSHamilton** [comment]
@Nirajn2311 Are you happy converting this from a draft - not completing the list of components (can get to them another time)?

**Comment by Nirajn2311** [comment]
Sure, I'll change it now. I'll push any final changes I have within the next 3 hours.

**Comment by ShaunSHamilton** [comment]
@Nirajn2311 Sorry about this, but I forgot to retarget this PR to `next` (done now). Would you mind rebasing one last time? _with next_

File: client/src/components/settings/Portfolio.tsx (hunk 3 above)
Comment (Nirajn2311) [reply client/src/components/settings/Portfolio.tsx]: I searched the whole repo, and the function isn't called anywhere.

File: client/src/components/settings/Portfolio.tsx (hunk 2 above)
Comment (Nirajn2311) [reply client/src/components/settings/Portfolio.tsx]: For now, I've left this function commented out.

**Comment by Nirajn2311** [comment]
@ShaunSHamilton I've rebased with `next` branch. Until `Spacer` and `BlockSaveButton`(found in "components/helpers/form") are migrated, there will be some errors.

File: client/src/components/settings/DeleteModal.tsx (hunk 1 above)
Comment (Nirajn2311) [reply client/src/components/settings/DeleteModal.tsx]: I've changed this to the one you suggested. But I haven't gone through the link Oliver sent yet. Will do that soon.

File: client/src/components/settings/Portfolio.tsx (hunk 3 above)
Comment (ShaunSHamilton) [reply client/src/components/settings/Portfolio.tsx]: It is called here:
```
  handleRemoveItem = id => {
    return this.setState(
      state => ({
        portfolio: state.portfolio.filter(p => p.id !== id)
      }),
      () => this.handleSubmit(mockEvent)
    );
  };
```
And used here (to not throw an error:
```
handleSubmit = e => {
    e.preventDefault();
```

**Comment by ShaunSHamilton** [comment]
> @ShaunSHamilton I've rebased with `next` branch. Until `Spacer` and `BlockSaveButton`(found in "components/helpers/form") are migrated, there will be some errors. If no one has picked those 2 files, I can create a PR for them quickly as I had done some changes to them in my local to get no TS errors.

Thanks, for that. We are prioritising a PR that refactors `Spacer` to correct the type error, but I see @thecodingaviator is working on `BlockSaveButton` (ref https://github.com/freeCodeCamp/freeCodeCamp/pull/42476/files#diff-0a8b2e829372f73c50d8a9177a38dd3678c03a525d7d9b4f7026b450dc7ad13a) - did not realise this was an issue...

File: client/src/components/settings/Portfolio.tsx (hunk 3 above)
Comment (Nirajn2311) [reply client/src/components/settings/Portfolio.tsx]: Umm I pasted that same function above in my comment here and yea those two were the only places which rely on it.

Comment (ShaunSHamilton) [reply client/src/components/settings/Portfolio.tsx]: I must be tired...sorry.

**Comment by thecodingaviator** [comment]
> > @ShaunSHamilton I've rebased with `next` branch. Until `Spacer` and `BlockSaveButton`(found in "components/helpers/form") are migrated, there will be some errors. If no one has picked those 2 files, I can create a PR for them quickly as I had done some changes to them in my local to get no TS errors.
> 
> Thanks, for that. We are prioritising a PR that refactors `Spacer` to correct the type error, but I see @thecodingaviator is working on `BlockSaveButton` (ref https://github.com/freeCodeCamp/freeCodeCamp/pull/42476/files#diff-0a8b2e829372f73c50d8a9177a38dd3678c03a525d7d9b4f7026b450dc7ad13a) - did not realise this was an issue...

@ShaunSHamilton #42476 should be ready to go. A quick review is required

File: client/src/components/settings/Portfolio.tsx (hunk 3 above)
Comment (Nirajn2311) [reply client/src/components/settings/Portfolio.tsx]: It's fine

**Pull request merged**